Feature/improvements

rburtnykx opened the pull request:
FIXED: issue when std output is redirected (sys.stdout.encoding is None)
ADDED: use compile base directory from gcov json; Here: https://gcc.gnu.org/onlinedocs/gcc/Invoking-Gcov.html#Invoking-Gcov 
{
  "**current_working_directory": current_working_directory**,
  "data_file": data_file,
...

ADDED: print statistic feature; 
ADDED: source existence check

RPGillespie6 commented:
Could you elaborate on what the `sys.stdout.encoding` issue is and how to reproduce?

I need to check if `current_working_directory` has been there since GCC 9.0, if so we can make this change. We should still allow the user to override it though, to retain backward compatibility.

rburtnykx commented:
Yes, sure. In case of redirected output, f.e. to file, `sys.stdout.encoding` is None. So error will be smth like next:

```
Traceback (most recent call last):
...
    intermediate_json = json.loads(line.decode(sys.stdout.encoding))
TypeError: decode() argument 1 must be string, not None
```
So to reproduce step is next:
`python fastcov.py > test.txt`
Script should run getGcovCoverage.

About GCC 9.0. Sure. Unfortunately, I was not able to check it for gcc 9.0, only 9.1+, but it was mentioned in all docs for gcc9.
About backward compatibility. How will be better to do that? I see 2 solutions:
1. adding additional cmdline parameter which will ingore -c, (like --gcov-determine-compile-dir) 
2. If -c is not set, use current_work_dir json field from gcov

RPGillespie6 commented:
Ah, I see. Good catch, I had not thought to try that. Thanks for adding the redirect (and other) test cases to `test/functional/run_all.sh`

> About backward compatibility. How will be better to do that? I see 2 solutions:

I think the second solution is better. If `-c` is not set, use gcov field. If `-c` is set, override gcov field with what was set.

I will take a closer look at this PR and see if I have any other suggestions, but so far I like these improvements.

rburtnykx commented:
Great, thanks. So i will definitely leave here this 2 items: redirect fix and base path using gcov output. Even if gcov has no that attribute in json, f.e. for 9.0.x version, it will behave the same after changes.

Other improvements are:
1. Exclude using glob. 
2. Dump short statistic at the end. That is smth like in lcov, sometimes useful to see that. If this is OK, i think i need to add also other info, like files, branches, etc.
3. Source existence check. 

Please, let me know if smth better to not include so i will remove that.

RPGillespie6 commented on fastcov.py:
```diff
@@ -368,6 +374,12 @@ def scanExclusionMarkers(fastcov_json, jobs, exclude_branches_sw, include_branch
+    logging.info("Checking if all sources exist")
+    for source in fastcov_json["sources"].keys():
+        if not os.path.exists(source):
+            logging.error("Cannot found '{}'".format(source))
```
Small typo, should be `Cannot find`

rburtnykx replied:
Fixed, thanks

rburtnykx commented:
Found and fixed a little issue with default file name. When --lcov parameter is set, default file name is coverage.json. Changed to  coverage.info if --lcov is set, coverage.json otherwise

RPGillespie6 commented:
Is this ready for merge now?

rburtnykx commented:
I think yes

RPGillespie6 commented:
Alright, I'm going to merge it. Thanks for contributing!

RPGillespie6 commented:
I'll probably make a follow on PR that bumps the fastcov version, adds a few more tests, and other preparations for `v1.8`

rburtnykx commented:
Awesome, thanks for taking my changes into branch.
